### PR TITLE
Worker improvements

### PR DESCRIPTION
**Pull request by Pessimistress**
- Support asynchronous `parse` in worker loaders
- Support using another loader (on main/another worker thread) inside a worker

### CODE REVIEW

File: modules/core/src/worker-utils/worker-thread.js [hunk 1]
```diff
@@ -1,13 +1,24 @@
...
 export default class WorkerThread {
-  constructor(workerSource, name = `web-worker-${count++}`) {
-    const url = getWorkerURL(workerSource);
+  constructor({source, name = `web-worker-${count++}`, parse}) {
+    const url = getWorkerURL(source);
     this.worker = new Worker(url, {name});
     this.name = name;
+
+    if (parse) {
+      const {port1, port2} = new MessageChannel();
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/core/src/worker-utils/worker-thread.js]: If we rely on `onmessage`, then there must be some sort of contract between `WorkerThread` and the worker code. This goes against the intention to separate `WorkerThread` from loaders specific logic.

File: modules/core/src/worker-utils/worker-thread.js [hunk 2]
```diff
@@ -28,4 +42,16 @@ export default class WorkerThread {
     this.worker.terminate();
     this.worker = null;
   }
+
+  _sanitizeOptions(options = {}) {
+    const result = {};
+    for (const key in options) {
+      let value = options[key];
+      if (isLoaderObject(value)) {
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/core/src/worker-utils/worker-thread.js]: How does it work if `DracoLoader` is not registered? Currently, the loader checks its `options` for whether an external decoder is available. It's impossible for it to access the global loader registry from a worker thread. If we send a parsing request to the main thread regardless, `parse` will throw if no appropriate loader is found.

File: modules/core/src/worker-utils/worker-thread.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ibgreen) [reply modules/core/src/worker-utils/worker-thread.js]: > then there must be some sort of contract between WorkerThread and the worker code.

I am not opposed to message channels so this part may not be worth a deep discussion but I am not sure I follow this. 

Two-way messaging between worker and main thread would seem to be a basic service that worker threads would provide. I.e. fact that a worker calls `onmessage` is not a contract.

Any contract would be in terms of the payload of the messages, which would be  defined by the calling layer. 

I do not see how using a different API makes a difference.

File: modules/core/src/worker-utils/worker-utils.js [hunk 3]
```diff
@@ -23,6 +23,14 @@ const workerURLCache = new Map();
 // Creates a URL from worker source that can be used to create `Worker` instances
 // Packages (and then caches) the result of `webworkify` as an "Object URL"
 export function getWorkerURL(workerSource) {
+  assert(typeof workerSource === 'string', 'worker source');
+
+  // url(./worker.js)
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/core/src/worker-utils/worker-utils.js]: Webpack bundles and serves the worker code into a file, and we can load it with

```js
new Worker('./json-loader.worker.js')
```

The existing worker logic only supports worker source, not url.

File: modules/core/src/worker-utils/worker-thread.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/core/src/worker-utils/worker-thread.js]: The current `WorkerThread.process` resolves whenever a message is received from the worker. If the worker is allowed to post various types of messages while decoding, then we need to define them. The workers that do not follow this system will no longer be supported.

[the remark this answers is not in the document]
Comment (ibgreen) [reply modules/core/src/worker-utils/worker-thread.js]: > The current WorkerThread.process resolves whenever a message is received from the worker.

> If the worker is allowed to post various types of messages while decoding, then we need to define them. 

Agreed, Ideally that messaging protocol is defined on the loaders.gl level, not the worker utils level.

To get around the "resolve on first message" issue, we could perhaps accept an `onMessage` callback to the `Worker` peek at each message and return and depending on its return value we resolve the `process` promise. Default callback would resolve on first message per current semantics.

File: modules/loader-utils/src/worker-utils/create-worker.js [hunk 4]
```diff
@@ -9,32 +9,65 @@ export default function createWorker(loader) {
     return;
   }
 
-  self.onmessage = evt => {
+  self.onmessage = async evt => {
```
Comment (ibgreen) [inline modules/loader-utils/src/worker-utils/create-worker.js]: Nit: I'd split this into a generic `createWorker` and a `createLoaderWorker` that adds the extra protocol processing.

The could be named differently (e.g createGenericWorker and createWorker) to not break the existing loader objects that depend on `createWorker`, but separating the basics and the "protocol" like you did on the calling side.

File: modules/core/src/worker-utils/worker-utils.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/core/src/worker-utils/worker-utils.js]: Yes, clarified in comment

File: modules/core/src/lib/parse-with-worker.js [hunk 5]
```diff
@@ -14,17 +15,44 @@ function getWorkerFarm(options = {}) {
...
+      return resolve(data);
+  }
+}
+
 /**
  * this function expects that the worker function sends certain messages,
  * this can be automated if the worker is wrapper by a call to createWorker in @loaders.gl/loader-utils.
  */
-export default async function parseWithWorker(workerSource, workerName, data, options) {
+export default function parseWithWorker(workerSource, workerName, data, options) {
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/core/src/lib/parse-with-worker.js]: How is it different in debugging without `async`?

File: modules/loader-utils/src/worker-utils/create-worker.js (hunk 4 above)
Comment (Pessimistress) [reply modules/loader-utils/src/worker-utils/create-worker.js]: There is no breaking change to the worker code. The only addition is a global `parse` function that communicates with the main thread.

File: modules/core/src/lib/parse-with-worker.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (ibgreen) [reply modules/core/src/lib/parse-with-worker.js]: > How is it different in debugging without async?

In this particular case, single stepping through the await will mean the async operation has completed in this call, as opposed to "a promise has been created" to be waited on somewhere else.

When there is a `.then`,`.catch` involved the difference is bigger, as there is no need to put breakpoints in different callbacks, and the amount of code clutter is significantly reduced.

File: modules/core/src/lib/parse-with-worker.js [hunk 6]
```diff
@@ -1,7 +1,8 @@
 import {toArrayBuffer} from '../javascript-utils/binary-utils';
 import WorkerFarm from '../worker-utils/worker-farm';
+import {parse} from './parse';
 
-import {removeNontransferableValues} from '../worker-utils/worker-utils';
+import {removeNontransferableValues, getTransferList} from '../worker-utils/worker-utils';
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/core/src/lib/parse-with-worker.js]: They are applied to different objects.

**Pull request merged**